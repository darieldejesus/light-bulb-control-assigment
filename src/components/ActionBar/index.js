import React from 'react';
import { Button, Txt, Box } from 'rendition';
import styled from 'styled-components';
import FaHome from 'react-icons/lib/fa/home';
import FaChevronLeft from 'react-icons/lib/fa/chevron-left';
import FaLightBulbO from 'react-icons/lib/fa/lightbulb-o';

const FaLightBulb = styled(FaLightBulbO)`
  margin: -0.2em 0.4em 0 0;
`;

export default () => (
  <Box pt={2} pb={[3, 4]} pl={3}>
    <Button py={2} px={0} m={0} quartenary square color="tertiary.main">
      <FaHome />
    </Button>
    <Button py={[0, 1, 2]} px={3} ml={2} quartenary color="tertiary.main">
      <FaChevronLeft />
      <Txt.span ml={2} fontSize={[0, 1]}>
        Back
      </Txt.span>
    </Button>
    <Txt.span ml={4} fontSize={[2, 3]} color="tertiary.main">
      <FaLightBulb />
      Lighting
    </Txt.span>
  </Box>
);
