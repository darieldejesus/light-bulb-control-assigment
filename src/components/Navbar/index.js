import React, { Component } from 'react';
import moment from 'moment';
import {
  Box,
  Txt,
  Flex,
} from 'rendition';

export default class Navbar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      date: this.getDate(),
      time: this.getTime(),
    };
  }

  componentDidMount() {
    this.interval = setInterval(() => this.updateDateTime(), 1000 * 60);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  /**
   * Get the current date i.e Friday 20, July, 2018.
   *
   * @returns {string} Current date.
   */
  getDate = () => moment().format('dddd D, MMMM, YYYY');

  /**
   * Get the current time i.e 10:54pm.
   *
   * @returns {string} Current time.
   */
  getTime = () => moment().format('h:mm a');

  updateDateTime = () => {
    this.setState({
      date: this.getDate(),
      time: this.getTime(),
    });
  };

  render() {
    const { date, time } = this.state;
    return (
      <Box>
        <Flex direction={['column', 'row']} bg="text.dark" alignItems="center">
          <Box width={1} pl={3} flex={2}>
            <Txt.p color="gray.light" fontSize={[0, 1]}>
              {date}
            </Txt.p>
          </Box>
          <Box width={1} flex={1}>
            <Txt.p color="gray.light" align="center" fontSize={[0, 1]}>
              {time}
            </Txt.p>
          </Box>
          <Flex width={1} pr={3} justifyContent="flex-end" flex={2}>
            <Txt.p color="gray.light" fontSize={[0, 1]}>
              Name Surname
            </Txt.p>
          </Flex>
        </Flex>
      </Box>
    );
  }
}
