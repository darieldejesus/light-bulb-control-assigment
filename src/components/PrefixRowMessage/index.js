import React from 'react';
import PropTypes from 'prop-types';
import { Txt } from 'rendition';

/**
 * Component to show a row with the given message in a table.
 *
 * @param {Object} obj - Main parameter.
 * @param {number} obj.colSpan - Column span.
 * @param {string} obj.message - Message to be shown.
 * @returns {Object} ReactElement.
 * */
export default function PrefixRowMessage({ colSpan, message }) {
  return (
    <tr>
      <td colSpan={colSpan}>
        <Txt.p align="center">
          {message}
        </Txt.p>
      </td>
    </tr>
  );
}

PrefixRowMessage.propTypes = {
  colSpan: PropTypes.number.isRequired,
  message: PropTypes.string.isRequired,
};
