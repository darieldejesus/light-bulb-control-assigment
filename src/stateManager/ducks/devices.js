import findIndex from 'lodash/findIndex';

// Actions
export const LOAD = 'resinio/devices/LOAD';
export const LOAD_SUCCESS = 'resinio/devices/LOAD_SUCCESS';
export const LOAD_FAIL = 'resinio/devices/LOAD_FAIL';
export const SELECT = 'resinio/devices/SELECT';
export const UPDATE = 'resinio/devices/UPDATE';
export const UPDATE_SUCCESS = 'resinio/devices/UPDATE_SUCCESS';
export const UPDATE_FAIL = 'resinio/devices/UPDATE_FAIL';
export const UPDATE_NAME = 'resinio/devices/UPDATE_NAME';
export const UPDATE_NAME_SUCCESS = 'resinio/devices/UPDATE_NAME_SUCCESS';
export const UPDATE_NAME_FAIL = 'resinio/devices/UPDATE_NAME_FAIL';
export const DISMISS = 'resinio/devices/DISMISS';

// Initial State
const initialState = {
  currentDevice: {},
  devices: [],
  isLoading: false,
  isUpdating: false,
  error: false,
  isUpdatingName: false,
};

/**
 * Update a device based on the response from the API.
 *
 * @param {Array} devices - List of devices in the state.
 * @param {number} deviceIndex - Index number where the old device is located.
 * @param {Object} newDevice - Updated device returned from the API.
 * @returns {Array} List of updated devices.
 * */
function updateDevice(devices, deviceIndex, newDevice) {
  return [
    ...devices.slice(0, deviceIndex),
    newDevice,
    ...devices.slice(deviceIndex + 1),
  ];
}

/**
 * Default Redux Reducer function.
 *
 * @param {Object} state - Current or default Redux state.
 * @param {Object} action - Object with the action type and data props.
 * @returns {Object} A new state based on given action.
 * */
export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case LOAD:
      return {
        ...state,
        isLoading: true,
      };
    case LOAD_SUCCESS: {
      const devices = action.result.data || [];
      return {
        ...state,
        devices,
        currentDevice: devices[0] || {},
        isLoading: false,
      };
    }
    case LOAD_FAIL:
      return {
        ...state,
        isLoading: false,
        error: 'Unable to load the devices from the server.',
      };
    case SELECT:
      return {
        ...state,
        currentDevice: action.payload,
      };
    case UPDATE:
      return {
        ...state,
        isUpdating: true,
      };
    case UPDATE_SUCCESS: {
      const updatedDevice = action.result.data;
      const deviceIndex = findIndex(state.devices, ['id', updatedDevice.id]);
      const devices = updateDevice(state.devices, deviceIndex, updatedDevice);
      return {
        ...state,
        isUpdating: false,
        devices,
        currentDevice: updatedDevice,
      };
    }
    case UPDATE_FAIL:
      return {
        ...state,
        isUpdating: false,
        error: 'Unable to update the selected light bulb.',
      };
    case UPDATE_NAME:
      return {
        ...state,
        isUpdatingName: true,
      };
    case UPDATE_NAME_SUCCESS: {
      const updatedDevice = action.result.data;
      const deviceIndex = findIndex(state.devices, ['id', updatedDevice.id]);
      const devices = updateDevice(state.devices, deviceIndex, updatedDevice);
      return {
        ...state,
        isUpdatingName: false,
        devices,
        currentDevice: updatedDevice,
      };
    }
    case UPDATE_NAME_FAIL:
      return {
        ...state,
        isUpdatingName: false,
        error: 'Unable to update the name of the selected light bulb.',
      };
    case DISMISS:
      return {
        ...state,
        error: false,
      };
    default:
      return state;
  }
}

/**
 * Action to get all the devices from the server/API.
 *
 * @returns {Object} Object with the required props to handle this action.
 * */
export function load() {
  return {
    types: [LOAD, LOAD_SUCCESS, LOAD_FAIL],
    promise: api => api.get('device'),
  };
}

/**
 * Action to set the current device when the user selects a row.
 *
 * @param {Object} device - Device object.
 * @returns {Object} Object with the required props to handle this action.
 * */
export function select(device) {
  return {
    type: SELECT,
    payload: device,
  };
}

/**
 * Action to toggle the light bulb status of the current device.
 *
 * @param {number} id - Device ID.
 * @param {Object} device - Properties to be updated.
 * @returns {Object} Object with the required props to handle this action.
 * */
export function toggleStatus(id, device) {
  const { active, brightness } = device;
  return {
    types: [UPDATE, UPDATE_SUCCESS, UPDATE_FAIL],
    promise: api => api.patch(`device/${id}`, {
      data: {
        ...device,
        active: !active,
        brightness: active ? 0 : brightness,
      },
    }),
  };
}

/**
 * Action to update the brightness of the current device.
 *
 * @param {number} id - Device ID.
 * @param {Object} device - Properties to be updated.
 * @returns {Object} Object with the required props to handle this action.
 * */
export function updateBrightness(id, device) {
  const { brightness, active } = device;
  return {
    types: [UPDATE, UPDATE_SUCCESS, UPDATE_FAIL],
    promise: api => api.patch(`device/${id}`, {
      data: {
        ...device,
        active: brightness === 0 ? false : active,
      },
    }),
  };
}

/**
 * Action to update the brightness of the current device.
 *
 * @param {number} id - Device ID.
 * @param {Object} device - Properties to be updated.
 * @returns {Object} Object with the required props to handle this action.
 * */
export function updateName(id, device) {
  return {
    types: [UPDATE_NAME, UPDATE_NAME_SUCCESS, UPDATE_NAME_FAIL],
    promise: api => api.patch(`device/${id}`, {
      data: device,
    }),
  };
}

/**
 * Action to dismiss the current error.
 *
 * @returns {Object} Object with the required props to handle this action.
 * */
export function dismissError() {
  return {
    type: DISMISS,
  };
}
