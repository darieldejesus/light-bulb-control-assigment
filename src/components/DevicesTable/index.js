import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Table } from 'rendition';
import debounce from 'lodash/debounce';
import styled from 'styled-components';
import RowCell from './RoomCell';

import Switch from '../ToggleSwitch';
import PrefixRowMessage from '../PrefixRowMessage';
import {
  load,
  select,
  toggleStatus,
} from '../../stateManager/ducks/devices';
import { colors } from '../../theme';

const mapStateToProps = ({ devices }) => ({
  ...devices,
});

const mapDispatchToProps = dispatch => ({
  loadDevices: () => dispatch(load()),
  selectDevice: device => dispatch(select(device)),
  toggleLightBulb: (id, { active, brightness }) => (
    dispatch(toggleStatus(id, { active, brightness }))
  ),
});

// Styles to highlight selected row.
const DevicesList = styled(Table)`
  a[data-key="${({ currentDeviceName }) => currentDeviceName}"] {
    background: ${colors.info.light};
  }
  a[data-key="${({ currentDeviceName }) => currentDeviceName}"]:first-child {
    box-shadow: inset 4px 0px ${colors.info.main};
  }
`;

export class DevicesTable extends Component {
  constructor(props) {
    super(props);
    this.columns = [
      {
        field: 'name',
        label: 'Room',
        sortable: true,
        render: (value, device) => (
          <RowCell
            value={value}
            currentDevice={device}
          />
        ),
      },
      {
        field: 'active',
        label: 'State',
        sortable: false,
        render: (value, device) => (
          <Switch onClick={this.handleOnToggleStatus(device)} active={value} />
        ),
      },
      {
        field: 'brightness',
        label: 'Brightness',
        sortable: false,
        render: value => `${value}%`,
      },
    ];
  }

  componentWillMount() {
    const { loadDevices } = this.props;
    loadDevices();
  }

  /**
   * Handles the click event of the Toggle Switch that turn off/on a light bulb.
   * It's defined as a debounce to prevent spam clicking.
   *
   * @param {Object} device - Device object.
   * @returns {Function} - A new debounce function with the desired behavior.
   */
  handleOnToggleStatus = device => debounce(() => {
    const { toggleLightBulb } = this.props;
    const { id, active, brightness } = device;
    toggleLightBulb(id, {
      active,
      brightness,
    });
  }, 150);

  /**
   * Handles the click event of the Devices Table when the user selects a row.
   *
   * @param {Object} device - Device object.
   * @returns {undefined} Void.
   */
  handleOnSelectDevice = (device) => {
    const { selectDevice } = this.props;
    selectDevice(device);
  };

  render() {
    const {
      isLoading,
      devices,
      currentDevice,
    } = this.props;
    const { name } = currentDevice;
    let prefixRow;
    if (isLoading) {
      prefixRow = (
        <PrefixRowMessage
          colSpan={this.columns.length}
          message="Loading devices..."
        />
      );
    }

    if (!prefixRow && !devices.length) {
      prefixRow = (
        <PrefixRowMessage
          colSpan={this.columns.length}
          message="No devices were found."
        />
      );
    }

    return (
      <DevicesList
        columns={this.columns}
        data={devices}
        rowKey="name"
        tbodyPrefix={prefixRow}
        onRowClick={this.handleOnSelectDevice}
        currentDeviceName={name}
        rowAnchorAttributes={{
          'data-device-name': name,
        }}
      />
    );
  }
}

DevicesTable.defaultProps = {
  isLoading: false,
  currentDevice: {
    name: '',
  },
};

DevicesTable.propTypes = {
  loadDevices: PropTypes.func.isRequired,
  toggleLightBulb: PropTypes.func.isRequired,
  selectDevice: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  currentDevice: PropTypes.shape({
    name: PropTypes.string,
  }),
  devices: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
    brightness: PropTypes.number.isRequired,
  })).isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(DevicesTable);
