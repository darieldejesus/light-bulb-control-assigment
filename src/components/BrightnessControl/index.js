import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import SliderSelectControl from '../SliderSelectControl';
import { updateBrightness } from '../../stateManager/ducks/devices';

/**
 * Define the state properties required by the component.
 *
 * @param {Object} obj - Main paramater.
 * @param {Object} obj.devices - Object with the current Devices state from Redux.
 * @returns {undefined} Void return.
 */
const mapStateToProps = ({ devices }) => ({
  ...(devices.currentDevice || {}),
});

/**
 * Define the actions to be accessed by the component.
 *
 * @param {Function} dispatch - Redux dispatcher.
 * @returns {undefined} Void return.
 */
const mapDispatchToProps = dispatch => ({
  updateDeviceBrightness: (id, { active, brightness }) => (
    dispatch(updateBrightness(id, { active, brightness }))
  ),
});

/**
 * Stateless Component to handle the Brightness control.
 *
 * @param {Object} obj - Props of the stateless component.
 * @param {number} obj.id - Id of the selected device.
 * @param {number} obj.brightness - Percentage number of the current device brightness.
 * @param {Function} obj.updateDeviceBrightness - Function to update the brightness.
 * @returns {Object} ReactElement object.
 */
export function BrightnessControl({
  id, active, brightness, updateDeviceBrightness,
}) {
  return (
    <SliderSelectControl
      strokeWidth={13}
      circleStartAngle={45}
      circleEndAngle={315}
      padding={50}
      radius={120}
      enabled={id > 0}
      value={brightness}
      onChange={brightnessPercentage => (
        updateDeviceBrightness(id, { active, brightness: brightnessPercentage })
      )}
    />
  );
}

BrightnessControl.defaultProps = {
  id: 0,
  brightness: 0,
  active: false,
};

BrightnessControl.propTypes = {
  updateDeviceBrightness: PropTypes.func.isRequired,
  brightness: PropTypes.number,
  active: PropTypes.bool,
  id: PropTypes.number,
};

export default connect(mapStateToProps, mapDispatchToProps)(BrightnessControl);
