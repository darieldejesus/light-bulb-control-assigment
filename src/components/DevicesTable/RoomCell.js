import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  Box,
  Txt,
  Button,
  Flex,
  Input,
} from 'rendition';
import FaPencil from 'react-icons/lib/fa/pencil';
import FaDiscard from 'react-icons/lib/fa/mail-reply';
import FaCheck from 'react-icons/lib/fa/floppy-o';
import FaUpdating from 'react-icons/lib/fa/spinner';
import styled from 'styled-components';

import { updateName } from '../../stateManager/ducks/devices';

const CellContainer = styled(Box)`
  position: relative;
`;

const ActionButton = styled(Button)`
  display: ${({ show }) => show ? 'inline-block' : 'none'}
`;

const EditButton = ActionButton.extend`
  position: absolute;
  left: -1.8em;
  top: 0.1em;
  
  &:hover, :active {
    color: unset;
  }
`;

const mapStateToProps = ({ devices }) => ({
  currentDevice: devices.currentDevice,
  isUpdating: devices.isUpdatingName,

});

const mapDispatchToProps = dispatch => ({
  updateDeviceName: (id, device) => dispatch(updateName(id, device)),
});

export class RoomCell extends Component {
  constructor(props) {
    super(props);
    this.state = {
      edit: false,
    };
  }

  /**
   * Prevent component to be re-rendered if isUpdating is active
   * but the component is not being edited.
   *
   * @param {Object} nextProps - Next component props.
   * @param {boolean} nextProps.isUpdating - Tells if a component is being updated.
   * @param {Object} nextState - Next component state.
   * @param {boolean} nextState.show - Tells if a component is selected.
   * @returns {boolean} False or True depending on the condition.
   */
  shouldComponentUpdate({ isUpdating }, { show }) {
    const { isUpdating: currentUpdating } = this.props;
    return !(isUpdating !== currentUpdating && !show);
  }

  static getDerivedStateFromProps({ value, currentDevice }, { name, edit }) {
    const { id, name: deviceName } = currentDevice;
    const show = value === deviceName;
    return {
      id,
      show,
      name: edit && show ? name : value,
      edit: edit && !show ? false : edit,
    };
  }

  handleOnToggleView = () => {
    this.setState(({ edit }) => ({
      edit: !edit,
    }));
  };

  handleOnChangeName = (e) => {
    this.setState({
      name: e.target.value,
    });
  };

  handleOnEditClick = () => {
    const { updateDeviceName } = this.props;
    const { id, name } = this.state;
    updateDeviceName(id, {
      name,
    });
  };

  render() {
    const { show, edit, name } = this.state;
    const { isUpdating } = this.props;
    const tooltip = edit && show
      ? 'Cancel edition'
      : 'Edit Room name';
    const icon = edit && show
      ? <FaDiscard />
      : <FaPencil />;

    return (
      <CellContainer>
        <Flex>
          <EditButton
            tooltip={tooltip}
            danger={edit}
            m={2}
            fontSize={[3, 2]}
            show={show}
            plaintext
            primary
            onClick={this.handleOnToggleView}
            iconElement={icon}
          />
          {edit && show ? ([
            <Input
              key="EditInput"
              value={name}
              autoFocus
              fontSize={0}
              width="8em"
              px={1}
              placeholder="Room name"
              onChange={this.handleOnChangeName}
            />,
            !isUpdating && (
              <ActionButton
                key="SaveChanges"
                tooltip="Save changes"
                success
                m={2}
                fontSize={[3, 2]}
                show={show}
                plaintext
                primary
                onClick={this.handleOnEditClick}
                iconElement={<FaCheck />}
              />
            ),
            isUpdating && (
              <ActionButton
                key="SaveChanges"
                success
                m={2}
                fontSize={[3, 4]}
                show={show}
                plaintext
                primary
                onClick={this.handleOnEditClick}
                iconElement={<FaUpdating />}
              />
            ),
          ]) : (
            <Txt.p>
              {name}
            </Txt.p>
          )}
        </Flex>
      </CellContainer>
    );
  }
}

RoomCell.propTypes = {
  value: PropTypes.string.isRequired, // eslint-disable-line
  currentDevice: PropTypes.shape({ // eslint-disable-line
    name: PropTypes.string.isRequired,
  }),
  isUpdating: PropTypes.bool.isRequired,
  updateDeviceName: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(RoomCell);
