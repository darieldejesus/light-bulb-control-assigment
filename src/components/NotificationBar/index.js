import React from 'react';
import { Alert } from 'rendition';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { dismissError } from '../../stateManager/ducks/devices';

const mapStateToProps = ({ devices }) => ({
  error: devices.error,
});

const mapDispatchToProps = dispatch => ({
  onDismiss: () => dispatch(dismissError()),
});

/**
 * Component to show a box with the error message.
 *
 * @param {Object} props - Component props.
 * @param {string} props.error - Error message.
 * @param {Function} props.onDismiss - Callback to dismiss the error.
 * @returns {Object} As ReactElement.
 */
function NotificationBar({ error, onDismiss }) {
  return (
    error && (
      <Alert
        mb={2}
        py={0}
        danger
        onDismiss={onDismiss}
      >
        { error }
      </Alert>
    )
  );
}

NotificationBar.defaultProps = {
  error: false,
};

NotificationBar.propTypes = {
  error: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.bool.isRequired,
  ]),
  onDismiss: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(NotificationBar);
