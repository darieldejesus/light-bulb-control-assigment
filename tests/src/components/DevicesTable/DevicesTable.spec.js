import React from 'react';
import { Provider as RenditionProvider } from 'rendition';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider as ReduxProvider } from 'react-redux';

import api from '../../../../src/helpers/ApiClient';
import clientMiddleware from '../../../../src/stateManager/middleware/client';
import devices from '../../../fixtures/devices';
import { DevicesTable } from '../../../../src/components/DevicesTable';
import theme from '../../../../src/theme';

jest.mock('lodash/debounce', () => jest.fn(fn => fn));

const middlewares = [
  clientMiddleware(api),
  thunk,
];
const mockStore = configureMockStore(middlewares);

describe('DevicesTable', () => {
  let props;

  beforeEach(() => {
    props = {
      loadDevices: () => {},
      toggleLightBulb: () => {},
      selectDevice: () => {},
      updateDeviceName: () => {},
      devices,
    };
  });

  test('Should render a table with the devices', () => {
    const wrapper = shallow(<DevicesTable {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should render a table even if devices but with a message', () => {
    props = {
      ...props,
      devices: [],
    };
    const wrapper = shallow(<DevicesTable {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should render a loading message if "isLoading" is active', () => {
    props = {
      ...props,
      devices: [],
      isLoading: true,
    };
    const wrapper = shallow(<DevicesTable {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should render an error message if "error" is defined', () => {
    props = {
      ...props,
      devices: [],
      error: true,
    };
    const wrapper = shallow(<DevicesTable {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should call loadDevices to fetch devices', () => {
    const loadDevices = jest.fn();
    props = {
      ...props,
      loadDevices,
      devices: [],
    };
    shallow(<DevicesTable {...props} />);
    expect(loadDevices).toBeCalled();
  });

  test('Should show formatted brightness as percentage numbers', () => {
    const brightness = 70;
    const formattedBrightness = `${brightness}%`;
    const currentDevice = {
      id: 2,
      name: 'Bedroom 01',
      active: false,
      brightness,
    };
    props = {
      ...props,
      devices: [currentDevice],
    };
    const store = mockStore({
      devices: {
        currentDevice,
        isUpdatingName: false,
      },
    });
    const wrapper = mount(
      <ReduxProvider store={store}>
        <RenditionProvider theme={theme}>
          <DevicesTable {...props} />
        </RenditionProvider>
      </ReduxProvider>,
    );
    expect(wrapper.find('a').at(2).text()).toEqual(formattedBrightness);
  });

  test('Should call toggleLightBulb with the device id and status', () => {
    const toggleLightBulb = jest.fn();
    const id = 2;
    const active = false;
    const brightness = 70;
    const device = {
      id,
      name: 'Bedroom 01',
      active,
      brightness,
    };
    props = {
      ...props,
      toggleLightBulb,
      devices: [device],
    };
    const wrapper = shallow(<DevicesTable {...props} />);
    wrapper.instance().handleOnToggleStatus(device)();
    expect(toggleLightBulb).toBeCalledWith(id, {
      active,
      brightness,
    });
  });

  test('Should call "selectDevice" with the selected device', () => {
    const selectDevice = jest.fn();
    const id = 2;
    const active = false;
    const device = {
      id,
      name: 'Bedroom 01',
      active,
      brightness: 70,
    };
    props = {
      ...props,
      selectDevice,
      devices: [device],
    };
    const wrapper = shallow(<DevicesTable {...props} />);
    wrapper.instance().handleOnSelectDevice(device);
    expect(selectDevice).toBeCalledWith(device);
  });
});
