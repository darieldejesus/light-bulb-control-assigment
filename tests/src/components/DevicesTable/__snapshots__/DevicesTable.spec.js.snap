// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevicesTable Should render a loading message if "isLoading" is active 1`] = `
<Styled(Table)
  columns={
    Array [
      Object {
        "field": "name",
        "label": "Room",
        "render": [Function],
        "sortable": true,
      },
      Object {
        "field": "active",
        "label": "State",
        "render": [Function],
        "sortable": false,
      },
      Object {
        "field": "brightness",
        "label": "Brightness",
        "render": [Function],
        "sortable": false,
      },
    ]
  }
  currentDeviceName=""
  data={Array []}
  onRowClick={[Function]}
  rowAnchorAttributes={
    Object {
      "data-device-name": "",
    }
  }
  rowKey="name"
  tbodyPrefix={
    <PrefixRowMessage
      colSpan={3}
      message="Loading devices..."
    />
  }
/>
`;

exports[`DevicesTable Should render a table even if devices but with a message 1`] = `
<Styled(Table)
  columns={
    Array [
      Object {
        "field": "name",
        "label": "Room",
        "render": [Function],
        "sortable": true,
      },
      Object {
        "field": "active",
        "label": "State",
        "render": [Function],
        "sortable": false,
      },
      Object {
        "field": "brightness",
        "label": "Brightness",
        "render": [Function],
        "sortable": false,
      },
    ]
  }
  currentDeviceName=""
  data={Array []}
  onRowClick={[Function]}
  rowAnchorAttributes={
    Object {
      "data-device-name": "",
    }
  }
  rowKey="name"
  tbodyPrefix={
    <PrefixRowMessage
      colSpan={3}
      message="No devices were found."
    />
  }
/>
`;

exports[`DevicesTable Should render a table with the devices 1`] = `
<Styled(Table)
  columns={
    Array [
      Object {
        "field": "name",
        "label": "Room",
        "render": [Function],
        "sortable": true,
      },
      Object {
        "field": "active",
        "label": "State",
        "render": [Function],
        "sortable": false,
      },
      Object {
        "field": "brightness",
        "label": "Brightness",
        "render": [Function],
        "sortable": false,
      },
    ]
  }
  currentDeviceName=""
  data={
    Array [
      Object {
        "active": true,
        "brightness": 50,
        "id": 1,
        "name": "Balcony",
      },
      Object {
        "active": false,
        "brightness": 70,
        "id": 2,
        "name": "Bedroom 01",
      },
      Object {
        "active": false,
        "brightness": 70,
        "id": 3,
        "name": "Bedroom 02",
      },
      Object {
        "active": true,
        "brightness": 100,
        "id": 4,
        "name": "Entrance",
      },
      Object {
        "active": true,
        "brightness": 100,
        "id": 5,
        "name": "Kitchen",
      },
      Object {
        "active": true,
        "brightness": 40,
        "id": 6,
        "name": "Living Room",
      },
      Object {
        "active": false,
        "brightness": 60,
        "id": 7,
        "name": "Master Bedroom",
      },
      Object {
        "active": false,
        "brightness": 100,
        "id": 8,
        "name": "Storage",
      },
    ]
  }
  onRowClick={[Function]}
  rowAnchorAttributes={
    Object {
      "data-device-name": "",
    }
  }
  rowKey="name"
/>
`;

exports[`DevicesTable Should render an error message if "error" is defined 1`] = `
<Styled(Table)
  columns={
    Array [
      Object {
        "field": "name",
        "label": "Room",
        "render": [Function],
        "sortable": true,
      },
      Object {
        "field": "active",
        "label": "State",
        "render": [Function],
        "sortable": false,
      },
      Object {
        "field": "brightness",
        "label": "Brightness",
        "render": [Function],
        "sortable": false,
      },
    ]
  }
  currentDeviceName=""
  data={Array []}
  onRowClick={[Function]}
  rowAnchorAttributes={
    Object {
      "data-device-name": "",
    }
  }
  rowKey="name"
  tbodyPrefix={
    <PrefixRowMessage
      colSpan={3}
      message="No devices were found."
    />
  }
/>
`;
